thrift/thrift-gen/tchannel-template.go

```diff
@@ -118,14 +120,49 @@ func (s *{{ .ServerStruct }}) Service() string {
 func (s *{{ .ServerStruct }}) Methods() []string {
 	return []string{
 		{{ range .Methods }}
-			"{{ .ThriftName }}",
+			"{{ .ThriftName }}", 
 		{{ end }}
 		{{ range .InheritedMethods }}
 			"{{ . }}",
 		{{ end }}
 	}
 }
 
+// RegisterInterceptors registers the provided interceptors with the server.
+func (s *{{ .ServerStruct }}) RegisterInterceptors(interceptors ...thrift.Interceptor) {
+	if s.interceptors == nil {
+		interceptorsLength := len(interceptors)
+		s.interceptors = make([]thrift.Interceptor, 0, interceptorsLength)
+	}
+
+	s.interceptors = append(s.interceptors, interceptors...)
+}
+
+func (s *{{ .ServerStruct }}) callInterceptorsPre(ctx {{ contextType }}, method string, args athrift.TStruct) error {
+	if s.interceptors == nil {
+		return nil
+	}
+	var firstErr error
+	for _, interceptor := range s.interceptors {
+		err := interceptor.Pre(ctx, method, args)
+		if err != nil && firstErr == nil {
+			firstErr = err
```

Comment:
So if you run into an error while executing a `Pre`, you still run through them all? That isn't a good idea.

Generally, interceptors semantics are:

- Stop at the first error in the 'pre' stage.
- Execute the 'post' of all interceptors that successfully executed and the one that failed **in reverse order**. That is, as soon as an interceptor fails, its post should be executed, then the post of the one before that, and so on. This allows for interceptors to perform clean up in the correct order.
- If there were no failures, simply run the 'post' of all interceptors in the reverse order. 

```
a.pre
    b.pre
        c.pre FAILED
            If c.pre had not failed, we would call the rest of the interceptors here
            followed by the actual handler
        c.post
    b.post
a.post
```